starboard/shared/libiamf/iamf_config_reader.cc

```diff
@@ -0,0 +1,287 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "starboard/shared/libiamf/iamf_config_reader.h"
+
+#include <string>
+
+#include "starboard/common/string.h"
+
+namespace starboard {
+namespace shared {
+namespace libiamf {
+
+namespace {
+// From
+// https://aomediacodec.github.io/iamf/v1.0.0-errata.html#obu-header-syntax.
+constexpr int kObuTypeCodecConfig = 0;
+constexpr int kObuTypeAudioElement = 1;
+constexpr int kObuTypeMixPresentation = 2;
+constexpr int kObuTypeSequenceHeader = 31;
+
+// From
+// https://aomediacodec.github.io/iamf/v1.0.0-errata.html#obu-codecconfig.
+constexpr int kFourccOpus = 0x4f707573;
+constexpr int kFourccMp4a = 0x6d703461;
+constexpr int kFourccFlac = 0x664c6143;
+constexpr int kFourccIpcm = 0x6970636d;
+
+constexpr int kLayoutTypeReserved = 1;
+constexpr int kLayoutTypeSpeaker = 2;
+constexpr int kLayoutTypeBinaural = 3;
+
+inline uint32_t ByteSwap(uint32_t x) {
+#if defined(COMPILER_MSVC)
+  return _byteswap_ulong(x);
+#else
+  return __builtin_bswap32(x);
+#endif
+}
+
+// Decodes an Leb128 value and stores it in |value|. Returns the number of bytes
+// read. Returns -1 on error.
+int ReadLeb128Value(const uint8_t* buf, uint32_t* value) {
+  SB_DCHECK(buf);
+  SB_DCHECK(value);
+  *value = 0;
+  bool error = true;
+  size_t i = 0;
+  for (; i < sizeof(uint32_t); ++i) {
+    uint8_t byte = buf[i];
+    *value |= ((byte & 0x7f) << (i * 7));
+    if (!(byte & 0x80)) {
+      error = false;
+      break;
+    }
+  }
+
+  if (error) {
+    return -1;
+  }
+  return i + 1;
+}
+
+int ReadString(const uint8_t* buf, std::string& value) {
+  SB_DCHECK(buf);
+  value = "";
+  value.reserve(128);
+
+  int bytes_read = 0;
+  bool error = true;
+  while (bytes_read < 128 && buf[bytes_read] != '\0') {
+    value.push_back(static_cast<char>(buf[bytes_read]));
+    bytes_read++;
+    error = false;
+  }
+
+  if (error) {
+    return -1;
+  }
+
+  return bytes_read;
+}
+}  // namespace
+
+bool IamfConfigReader::ResetAndRead(scoped_refptr<InputBuffer> input_buffer) {
+  Reset();
+  return Read(input_buffer);
+}
+
+void IamfConfigReader::Reset() {
+  buffer_head_ = 0;
+  mix_presentation_id_ = std::optional<uint32_t>();
+  config_size_ = 0;
+  data_size_ = 0;
+  sample_rate_ = 0;
+  samples_per_buffer_ = 0;
+  sample_size_ = 0;
+}
+
+bool IamfConfigReader::Read(scoped_refptr<InputBuffer> input_buffer) {
+  const uint8_t* buf = input_buffer->data();
+  SB_DCHECK(buf);
+
+  bool completed_parsing = false;
```

Comment:
This seems to be indicating that the OBU is the last one, consider naming it something like "is_last_obu", etc..